### PR TITLE
WIP: React-Redux Curriculum Migration

### PR DESCRIPTION
**Pull request by bonham000**
@QuincyLarson @BerkeleyTrue this branch establishes compatibility between the [alpha version of the freeCodeCamp React/Redux Curriculum](https://github.com/bonham000/fcc-react-tests-module) and the core freeCodeCamp app.

**Note:** This is a redo of previous PR because of issues with that branch. I've preserved my commit history here as a record of the sequence of changes I made.

I've figured out how to reproduce the testing environment we used in the Alpha version of these challenges within the freeCodeCamp app, and this PR successfully demonstrates the implementation of several specific challenges. The challenges I selected as a whole encompass all of the functionality we rely on, so this effectively paves the way for all of the challenges to be migrated over.

I'm aware that @BerkeleyTrue has a large, unmerged PR related to the new app layout (and possibly the React challenges as well?). There will possibly be merge conflicts between these two branches, so I'll defer to you guys to decide how to proceed from here.

Here, @no-stack-dub-sack and I are ready to go and begin migrating all the challenges over. That will be a pretty tedious process, but I think that is all that is left now (with the exception of QA) after a consensus is reached on how to integrate these changes into the app. We could either, work on this branch until all the challenges are included here, or, merge this support in and then just start making smaller PRs that add in the actual challenges.

### Main Notes:

- I provided the React/Redux challenges with a new `challengeType`, `8`, in order to differentiate them from the other challenges. For the most part, they are treated like `js` challenges, but this allows us to throw a new error toast if code transpilation fails and potentially pipe these challenges through a separate transformation process (to provide the `stage-2` preset). I can expand on this later if needed.
- The challenges rely on many dependencies in order to work, e.g. React, Redux, React-Redux packages. I provided these packages via imports directly within the app onto the document which runs the tests. The Enzyme methods we utilize are provide in a simliar manner.
- I don't know how the unique IDs are generated from the challenges (in the JSON files). I was just changing random characters in the IDs to get the challenges to seed the database.
- I also added an `npm restart` script which reseeds the database and restarts the server in one command to make the process of updating challenges a little more streamlined.
- The automated tests that Gulp performs should be compatible with many of the React challenges now. However, there were some problems related to enabling class property transforms with Babel and with running certain Enzyme tests which are not yet resolved, however these should only affect a few challenges.

I have some more notes, but they are more related to get the tests for the challenges working correctly. If we move forward with migrating these I'd be happy to create an issue to track that process and document the other notes I have.

### Demo:

react

react-redux

***

#### Pre-Submission Checklist
- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [ ] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit): commits unsquashed for review.
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change
- [x] New feature: this branch establishes the functionality to run the React/Redux challenges in the freeCodeCamp app.

#### Checklist:
- [x] Tested changes locally.

### CODE REVIEW

**Comment by QuincyLarson** [comment]
@bonham000 Awesome - thanks for submitting this PR. As you noted, @BerkeleyTrue is in the midst of a massive PR that will address a lot of these things. There may be some duplicate work between your PR and his. So I'll leave it up to him as to how we proceed here.

We do want to start importing the React and Redux challenges. I just want to make sure they're in the right format and they'll work properly with @BerkeleyTrue's code as well, so your scarce time porting these challenges put to productive use.

**Comment by BerkeleyTrue** [comment]
@bonham000 Many things are going to change with the new layout coming through. I'd hold off on making architectural changes.

**Comment by BerkeleyTrue** [comment]
@bonham000 Also, instead of editing camperbots automated messages, just add another comment to the thread.

File: client/rechallenge/transformers.js
Comment (BerkeleyTrue) [inline client/rechallenge/transformers.js]: I'd prefer we not teach anything that has not made it into the spec. class properties is not yet part of the spec and not specifically related to react.

File: client/sagas/build-challenge-epic.js
Comment (BerkeleyTrue) [inline client/sagas/build-challenge-epic.js]: We don't want to add paths specify to challenge types. The next iteration (modern) should handle everything without requiring specific paths. Otherwise adding new curriculum will get out of hand quickly.

File: client/sagas/err-saga.js
Comment (BerkeleyTrue) [inline client/sagas/err-saga.js]: These should be caught right at the compiler, this file is a generic error handler

File: client/sagas/frame-epic.js
Comment (BerkeleyTrue) [inline client/sagas/frame-epic.js]: We don't want to bind react challenges to the react used by the app. This couples our app to the challenges and means we won't be able to upgrade the app without first considering whether all the challenges are not effected.

We have a mechanism to import required libraries for challenges built into the rechallenge framework. This should be where we import libraries for all challenges.

File: common/app/routes/challenges/redux/selectors.js
Comment (BerkeleyTrue) [inline common/app/routes/challenges/redux/selectors.js]: `challengeType` is deprecated. We should be using the modern view for all new challenge types where applicable. This are challenges with `type: 'modern'`.

File: common/app/routes/challenges/redux/step-challenge-epic.test.js
Comment (BerkeleyTrue) [inline common/app/routes/challenges/redux/step-challenge-epic.test.js]: Always keep PR's scoped to the specific change you are introducing. I understand why this change is here, but it adds a lot of noise to the PR that makes it hard to parse your specific changes.

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

File: client/sagas/err-saga.js
Comment (bonham000) [inline client/sagas/err-saga.js]: The errors were being caught but the issue I saw was the feedback to the user wasn't clear, just "Something went wrong" so I add this saga specifically to catch transpilation errors from the React challenges to indicate to the user that the problem was that their code would not transpile. In the alpha version, each challenges has a specific test to see if the code transpiles or not.

File: client/sagas/build-challenge-epic.js
Comment (bonham000) [inline client/sagas/build-challenge-epic.js]: Sounds good!

File: client/rechallenge/transformers.js
Comment (bonham000) [inline client/rechallenge/transformers.js]: So this was unclear when we wrote the challenges which way we should go. The `class transforms` is for writing a class method with an arrow function to have the advantage of implicit `this` binding. Peter and I decided to add a challenge on it just to demonstrate the syntax (because we found it pretty common in other resources) but make it clear in the challenge that this is a future JavaScript version and the feature may not be added. We can remove it, but that was the motivation behind adding it originally.

**Pull request closed without merging**
